binding/react/src/use_rhino.ts

```diff
@@ -9,276 +9,108 @@
-  };
 
-  const pushToTalk = (): boolean => {
-    if (webVoiceProcessor !== null && rhinoWorker !== null) {
-      setIsTalking(true);
-      rhinoWorker.postMessage({ command: 'resume' });
-      return true;
-    }
-    return false;
-  };
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
+  const inferenceCallbackRef = useRef((inference: RhinoInference) => {});
+  const inferenceCallbackInner = useRef((inference: RhinoInference) => {
+    inferenceCallbackRef.current(inference);
+  });
 
-  /** Refresh the inference callback when it changes (avoid stale closure) */
   useEffect(() => {
-    if (typeof inferenceCallback !== 'function') {
-      // eslint-disable-next-line no-console
-      console.warn('useRhino: inferenceCallback is not a function');
-    }
-
-    if (rhinoWorker !== null) {
-      rhinoWorker.onmessage = (
-        msg: MessageEvent<RhinoWorkerResponse>
-      ): void => {
-        switch (msg.data.command) {
-          case 'rhn-inference':
-            setIsTalking(false);
-            rhinoWorker.postMessage({ command: 'pause' });
-            // We know this inference isFinalized, so assert to more specific type
-            inferenceCallback(msg.data.inference as RhinoInference);
-            break;
-          case 'rhn-error':
-            setIsError(true);
-            setErrorMessage(msg.data.error.toString());
-            break;
-          case 'rhn-info':
-            setContextInfo(msg.data.info);
-            break;
-          default:
-            break;
+    inferenceCallbackRef.current = (inference: RhinoInference): void => {
+      if (inference && inference.isFinalized) {
+        setIsListening(false);
+        if (rhino) {
+          WebVoiceProcessor.unsubscribe(rhino);
         }
-      };
-    }
+        inferenceCallback(inference);
+      }
+    };
   }, [inferenceCallback]);
 
-  /** Startup (and cleanup) Rhino */
-  useEffect(() => {
-    // Reset hook state, since we're rebooting Rhino and WebVoiceProcessor
-    setIsTalking(rhinoHookArgs?.isTalking === true);
-    setIsListening(false);
-    setIsLoaded(false);
-    setIsError(false);
-    setErrorMessage(null);
-
-    if (rhinoWorkerFactory === null || rhinoWorkerFactory === undefined) {
-      return (): void => {
-        /* NOOP */
-      };
+  const start = useCallback(async (): Promise<void> => {
+    try {
+      if (!rhino) {
+        const handle = await RhinoWorker.create(
+          accessKey,
+          context,
+          inferenceCallbackInner.current,
+          model,
+          options
+        );
+        setRhino(handle);
+        setContextInfo(handle.contextInfo);
+      }
+    } catch (e: any) {
+      setError(e.toString());
     }
+  }, [rhino, accessKey, context, model, options]);
 
-    if (rhinoHookArgs === null || rhinoHookArgs === undefined) {
-      return (): void => {
-        /* NOOP */
-      };
+  const pushToTalk = useCallback(async (): Promise<void> => {
```

Comment:
is it ok if someone uses pushToTalk twice before finalizing the first inference? if I'm not mistaken we subscribe the engine twice and that could cause some issues, right?

Reply:
Excellent point, it will cause issues so I added in a check